sys/crypto/modes/ccm.c

```diff
@@ -105,21 +105,35 @@ int ccm_compute_adata_mac(cipher_t* cipher, uint8_t* auth_data,
     if (auth_data_len > 0) {
         int len;
 
-        /* 16 octet block size + max. 10 len encoding  */
-        uint8_t auth_data_encoded[26], len_encoding = 0;
+        uint8_t auth_data_encoded[16] = {0}, len_encoding = 0;
 
-        if ( auth_data_len < (((uint32_t) 2) << 16)) {       /* length (0x0001 ... 0xFEFF)  */
+        if (auth_data_len < (((uint32_t) 1) << 16) - (1 << 8)) {
+            /* length (0x0001 ... 0xFEFF)  */
             len_encoding = 2;
 
             auth_data_encoded[1] = auth_data_len & 0xFF;
             auth_data_encoded[0] = (auth_data_len >> 8) & 0xFF;
         } else {
-            DEBUG("UNSUPPORTED Adata length\n");
+            DEBUG("UNSUPPORTED Adata length: %u\n", auth_data_len);
             return -1;
         }
 
-        memcpy(auth_data_encoded + len_encoding, auth_data, auth_data_len);
-        len = ccm_compute_cbc_mac(cipher, X1, auth_data_encoded, auth_data_len + len_encoding, X1);
+        int block_size = cipher_get_block_size(cipher);
+        size_t first_block_adata_len = block_size - len_encoding;
+        if (first_block_adata_len > auth_data_len)
+            first_block_adata_len = auth_data_len;
+        /* Create the first block of encoded auth data */
+        memcpy(auth_data_encoded + len_encoding, auth_data,
+               first_block_adata_len);
+        /* Compute CBC-MAC for the first block */
+        len = ccm_compute_cbc_mac(cipher, X1, auth_data_encoded, block_size, X1);
```

Comment:
It took me some time to understand why you are doing all this and not only computing it for the first two bytes but this is not my expertise field. So if I understood correctly `ccm_compute_cbc_mac` should only be called with full blocks, except for the last call that can have less. My review is based on this.

Using `block_size` here, required initializing `auth_data_encoded` to zero value to mimic `ccm_compute_cbc_mac` behavior (without a comment about it ;)). If you used `len_encoding + first_block_adata_len` here it would remove the need to initialize `auth_data_encoded`.
If `auth_data_len` was small enough so `len + auth_data_len` would be smaller than a `block_size` you would also not need the final `ccm_compute_cbc_mac` anyway so it should work as expected.

Also a reason for my hard time understanding also was how `first_block_adata_len` is set, it should fit in `block_size` but its minimized to `first_block_adata_len`.
I would rather set it to `auth_data_len` and trim it down to `block_size - len_encoding` if too big.
This way, this `first_block_adata_len` is initially linked to `auth_data_len` and minimized to fit for the `memcpy`.

It would keep the current behavior without computing empty bytes on small `auth_data_len`.

Either than these organization issues, this part of thee code looks to do what expected, but did not tested it.